### ISSUE TITLE
key colindex not found

### ISSUE THREAD

**Issue by Datseris** [JuliaData/DataFrames.jl#1456]
```
julia> typeof(file)
DataFrame

julia> fieldnames(typeof(file))
(:columns, :colindex)

julia> file[1]
149-element Array{Union{Missing, Int64},1}:
 0
  missing
  missing
  missing
  missing
  missing
  missing
  missing
 1
  missing
  missing
  missing
  missing
  missing
 0
  missing
  missing

julia> file.colindex
ERROR: KeyError: key :colindex not found
Stacktrace:
 [1] getindex at .\dict.jl:478 [inlined]
 [2] getindex at C:\Users\datseris\.julia\packages\DataFrames\LwtB\src\other\index.jl:123 [inlined]
 [3] getindex(::DataFrame, ::Symbol) at C:\Users\datseris\.julia\packages\DataFrames\LwtB\src\dataframe\dataframe.jl:258
 [4] getproperty(::DataFrame, ::Symbol) at C:\Users\datseris\.julia\packages\DataFrames\LwtB\src\abstractdataframe\abstractdataframe.jl:240

...

on julia 0.7

**Comment by nalimilan** [comment]
That's expected since we now override `getproperty`. Use `getfield` or `DataFrames.colindex(df)` instead. Or better, don't access it at all. ;-) Why do you want to access that internal field?

**Comment by Datseris** [comment]
Unfortunately, I do not know why. I had an old code for `DataFrames v0.10.0` I had to re-use, and at some point I was doing
```
# Read the file with all the data
file = readtable("rawdata/swing.valid.nocommas.csv";  separator=';')
ids = file[:id] # subject's id number
# But only keep the data about the recordings:
file = file[31:end-3]

# All the possible combinations as strings:
possibilities = String.(file.colindex.names)
```
don't really remember what it did... :D 

Is this issue documented? If yes the please close it. If not please keep it open until it is documented.

**Comment by Datseris** [comment]
What are you talking about?... This was a perfectly valid behavior that broke on current version. To the best of my knowledge there is no such thing as "private fields" in Julia. All fields of a `struct` are as "public" as they can get. 

```
julia> a = 1//4
1//4

julia> a.num
1
```

> It's not documented

So you think changes should never be documented? I couldn't find any `CHANGELOG` file in the main repo, nor a section "breaking changes" or whatever in the latest documentation page.

So how can I upgrade my code to DataFrames? By guessing / posting issues all the time?

**Comment by KristofferC** [comment]
> To the best of my knowledge there is no such thing as "private fields" in Julia. All fields of a struct are as "public" as they can get.

Private doesn't mean non-accessible. It means that it is an implementation detail.

> So how can I upgrade my code to DataFrames? By guessing / posting issues all the time?

By not relying on internal things like the names of fields.

**Comment by Datseris** [comment]
> By not relying on internal things like the names of fields.

You are right. I do want to update and I do want to write better code. I don't know what `file.colindex.names` does without running it. Also without any list of changes / warning I also can't know with what to replace it. Also, there is no documentation of `names` in the "latest" page: http://juliadata.github.io/DataFrames.jl/latest/lib/functions.html#Basics-1

Aaaalso, there is no documentation of `colindex`, which is now something unaccessible: http://juliadata.github.io/DataFrames.jl/latest/search.html?q=colindex 

also, there is no other place where `names` is in the documentation besides the broken page I've linked above.

By the way, the more people that try to justify the idea "yes, not listing changes is a good thing", the worse it seems to me.

**Comment by KristofferC** [comment]
The documentation block seems broken.

> By the way, the more people that try to justify the idea "yes, not listing changes is a good thing", the worse it seems to me.

You seem to be missing the distinction between what is considered implementation and what is considered the public facing API. Changes in the public facing API should indeed be documented, but listing every single change ("the internal global variable `__ABC` was renamed `__ADB`") is just noise.

**Comment by Datseris** [comment]
> but listing every single change ("the internal global variable `__ABC` was renamed `__ADB`") is just noise.

I totally agree. I disagree that accessing first level field falls in this category. We are both free to disagree, so there is no reason to continue this "leading-to-nowhere" conversation.

**Comment by KristofferC** [comment]
> I disagree that accessing first level field falls in this category. 

It does though, and if you keep using it in other packages your code will keep breaking. This cannot be agreed/disagreed on a personal basis, it is a convention used throughout Julia and by not following it you will get sad (like here).

**Comment by nalimilan** [comment]
> You are right. I do want to update and I do want to write better code. I don't know what file.colindex.names does without running it. Also without any list of changes / warning I also can't know with what to replace it. Also, there is no documentation of names in the "latest" page: http://juliadata.github.io/DataFrames.jl/latest/lib/functions.html#Basics-1

Yes, we need more docstrings, and we're well aware of it, but complaining won't make them appear by miracle.

**Comment by Liso77** [comment]
> That's expected since we now override getproperty. Use getfield or DataFrames.colindex(df) instead. Or better, don't access it at all. ;-) Why do you want to access that internal field?

Maybe I miss something. Why not to rewrite 
https://github.com/JuliaData/DataFrames.jl/blob/master/src/abstractdataframe/abstractdataframe.jl#L240
from
```julia
Base.getproperty(df::AbstractDataFrame, col_ind::Symbol) = getindex(df, col_ind)
```
to 
```julia
Base.getproperty(df::AbstractDataFrame, col_ind::Symbol) = getfield(df, col_ind) 
```
?

**Comment by nalimilan** [comment]
Because the point is to make `df.col` work as a shorthand for `df[:col]`.

**Comment by Liso77** [comment]
Thanks it is clear! :)

But if getproperty is less performant anyway is it problem to use something like:
```julia
Base.getproperty(df::AbstractDataFrame, sym::Symbol) = 
    sym ∈ names(df) ? getindex(df, sym) : getfield(df, sym)
```
?

**Comment by nalimilan** [comment]
The problem isn't performance, is that it's perfectly correct to have a column named `:colindex`, and we shouldn't return the private field in that case.

**Comment by Liso77** [comment]
And proposal above would return `getindex` and not `getfield` in that case. 

(BTW Python pandas.DataFrame would return getfield (at least in 0.23.3 version))
